### PR TITLE
feat: add directory replication support for multi-tenant databases

### LINKED ISSUE #42: Replicate whole directories
This is a great project! Would it be possible to configure replication for whole directories with multiple database files, with arbitrary number of nested directories?

### PR DESCRIPTION
**Pull request by corylanou**
## Summary

Adds directory replication support for multi-tenant applications where each tenant has their own SQLite database file.

Closes #42

## Configuration

```yaml
dbs:
  - directory: /var/lib/tenants
    pattern: "*.db"
    recursive: true
    replica:
      type: s3
      bucket: backups
      path: tenants
```

## Replica Path Behavior

Each database gets a unique replica path by appending its relative path from the directory root:

**Example**:
```
directory: /var/lib/databases
replica path: backups/prod

Results:
  /var/lib/databases/tenant1.db        → backups/prod/tenant1.db/ltx/...
  /var/lib/databases/team-a/db2.db     → backups/prod/team-a/db2.db/ltx/...
```

This ensures isolated storage per database with no collision risk.

## Features

- Scan directories for SQLite databases with configurable patterns (`*.db`, `*.sqlite`, etc.)
- Recursive directory scanning
- SQLite header validation (prevents replicating non-SQLite files)
- Mix directory and single-database configs
- Preserves directory structure in replica paths
- Works with all storage backends (S3, GCS, ABS, SFTP, WebDAV, File, NATS)

## Testing

- Comprehensive test coverage for path uniqueness
- Tests for subdirectory handling, duplicate filenames, special characters
- All existing tests passing

### CODE REVIEW

**Comment by corylanou** [comment]
## Update: Critical P0 Bug Fixed ✅

### Issue Identified During Review

Found a **critical data corruption bug** where all databases in a directory were sharing the same replica path, causing them to overwrite each other's LTX files.

### Root Cause

The original implementation performed a shallow copy of `DBConfig`, so all database instances shared the same `ReplicaConfig` pointer and wrote to the same replica path.

### Fix Applied (Commits 6f4ca03 & 00cdb7a)

**Solution**: Deep copy replica configs and append database-relative path to ensure uniqueness.

**Path Behavior**:
```yaml
directory: /var/lib/databases
replica:
  path: backups/prod

Results:
  /var/lib/databases/db1.db         → backups/prod/db1.db/ltx/...
  /var/lib/databases/team-a/db2.db  → backups/prod/team-a/db2.db/ltx/...
  /var/lib/databases/dir/db3.db     → backups/prod/dir/db3.db/ltx/...
```

Directory structure is preserved, eliminating collision risk even with duplicate filenames.

### Implementation Details

**Helper Functions** (refactored for clarity):
- `cloneReplicaConfigWithRelativePath()`: Deep copies replica config and appends relative path
- `appendRelativePathToURL()`: Handles URL-based replica configurations

**Type-Specific Handling**:
- File backend: Uses OS-specific paths (`filepath.Join`)
- Cloud backends: Uses forward slashes (`path.Join`)
- URL configs: Modifies URL path component directly

### Testing

**New Tests Added** (6 comprehensive test cases):
- ✅ Unique paths for multiple databases
- ✅ Subdirectory path preservation
- ✅ Duplicate filename collision prevention
- ✅ Special characters in filenames
- ✅ Empty base path edge case
- ✅ Deprecated 'replicas' array field
- ✅ URL-based configurations (both replica and replicas)

**All Tests Passing**:
```
✅ All 8 new directory replication tests
✅ All existing tests
✅ go vet clean
✅ staticcheck clean
✅ Build successful
```

### Breaking Change Notice

Since the original implementation causes data corruption, any existing users of directory replication will need to:
1. Delete corrupted replicas
2. Update to this fixed version
3. Restart replication (will use new path structure)

### Current PR Status

- **Rebased** onto latest main (22b5ce1)
- **P0 bug** completely resolved
- **Code refactored** for clarity and maintainability
- **Comprehensively tested** with edge cases
- **Documentation updated** with path behavior examples
- **Ready for merge** ✅

### Commits in PR

1. `5d204f8` - Rebase onto main with conflict resolution
2. `6f4ca03` - Fix P0 replica path collision bug
3. `00cdb7a` - Refactor with helper functions for better code organization

---

**Note**: This fix is critical. The feature cannot be safely used without it, as it will cause silent data corruption.

**Comment by corylanou** [comment]
# Directory Replication Field Report

## Environment

```text
litestream binary: ./litestream (go build ./cmd/litestream)
litestream-test:   ./bin/litestream-test (go build -o bin/litestream-test ./cmd/litestream-test)
MinIO container:   docker run -d --name litestream-minio -p 9500:9000 -p 9600:9090 \
                       -e MINIO_ROOT_USER=litestream \
                       -e MINIO_ROOT_PASSWORD=litestream123 \
                       minio/minio server /data --console-address :9090
mc client:         /tmp/mc (curl download)
Bucket:            litestream-directory-tests
```

MinIO credentials were also exported into the Litestream config using the `access-key-id` and `secret-access-key` fields.

## Local Directory Layout

SQLite databases populated under `/tmp/litestream-dir-test-hJj7VN`:

```text
/tmp/litestream-dir-test-hJj7VN
├─ db1.db
├─ team-a/
│  └─ db2.db
├─ team-b/
│  └─ east/
│     └─ db3.db
└─ shared/
   └─ db4.db
```

Each database was created with ~5 MB of data via:

```bash
./bin/litestream-test populate -db <path> -target-size 5MB
```

## Litestream Configuration

`/tmp/litestream-dir-test-config.yml`

```yaml
logging:
  level: info
  type: text

dbs:
  - directory: /tmp/litestream-dir-test-hJj7VN
    pattern: "*.db"
    recursive: true
    replica:
      type: s3
      bucket: litestream-directory-tests
      path: directory-run
      region: us-east-1
      endpoint: http://127.0.0.1:9500
      force-path-style: true
      skip-verify: true
      access-key-id: litestream
      secret-access-key: litestream123
```

## Replication Runs

Spin up Litestream using the config above:

```bash
./litestream replicate -config /tmp/litestream-dir-test-config.yml
```

Key log excerpts (`/tmp/litestream-dir-test-run.log` etc.):

```text
time=… level=INFO msg="found databases in directory" directory=/tmp/litestream-dir-test-hJj7VN count=4
time=… level=INFO msg="replicating to" path=directory-run/db1.db …
time=… level=INFO msg="replicating to" path=directory-run/team-a/db2.db …
time=… level=INFO msg="replicating to" path=directory-run/team-b/east/db3.db …
time=… level=INFO msg="replicating to" path=directory-run/shared/db4.db …
```

Expected warnings seen when stopping the process early:

- `compaction failed … page size not initialized yet` (harmless startup race)
- `failed to close database … transaction has already been committed or rolled back` (signal during shutdown)

## Load Generation

While replication ran, each database received write/read load from `litestream-test`:

```bash
for db in \
  "/tmp/litestream-dir-test-hJj7VN/db1.db" \
  "/tmp/litestream-dir-test-hJj7VN/team-a/db2.db" \
  "/tmp/litestream-dir-test-hJj7VN/team-b/east/db3.db" \
  "/tmp/litestream-dir-test-hJj7VN/shared/db4.db"; do
  ./bin/litestream-test load \
    -db "$db" \
    -duration 6s \
    -write-rate 40 \
    -workers 2 \
    -read-ratio 0.1
done
```

Results (aggregated from `/tmp/litestream-dir-test-run-load*.log`):

| Database                      | Writes/sec | Reads/sec | Errors |
|------------------------------|-----------:|----------:|-------:|
| db1.db                       | 35–44      | 3–6       | 0      |
| team-a/db2.db                | 35–44      | 3–6       | 0      |
| team-b/east/db3.db           | 34–45      | 4–6       | 0      |
| shared/db4.db                | 35–45      | 4–5       | 0      |

(All runs lasted 6–8 seconds with two workers and 1 KB payloads.)

## Remote Bucket Snapshot

`/tmp/mc tree --files litestream/litestream-directory-tests/directory-run`

```text
directory-run/
├─ db1.db/
│  ├─ 0000/0000000000000002-000000000000000a.ltx
│  └─ 0001/{0000000000000001-0000000000000001,…,0000000000000002-000000000000000a}.ltx
├─ team-a/db2.db/
│  ├─ 0000/0000000000000002-000000000000000a.ltx
│  ├─ 0001/0000000000000001-000000000000000a.ltx
│  └─ 0009/0000000000000001-0000000000000001.ltx
├─ team-b/east/db3.db/
│  ├─ 0000/…000a.ltx
│  ├─ 0001/0000000000000001-0000000000000007.ltx
│  └─ 0009/0000000000000001-0000000000000001.ltx
└─ shared/db4.db/
   ├─ 0000/…000a.ltx
   ├─ 0001/0000000000000001-0000000000000001.ltx
   └─ 0009/0000000000000001-0000000000000001.ltx
```

Each database replicates into a unique prefix derived from its relative path, confirming the directory-expansion fix works under active load.

## Artifacts

- Replication logs: `/tmp/litestream-dir-test-run*.log`
- Load logs: `/tmp/litestream-dir-test-run-load*.log`
- Config: `/tmp/litestream-dir-test-config.yml`
- Remote tree snapshot: `/tmp/mc tree --files …/directory-run`

## Cleanup Checklist

```bash
docker rm -f litestream-minio
rm -rf /tmp/litestream-dir-test-hJj7VN /tmp/litestream-dir-test-*.log /tmp/litestream-dir-test-config.yml
rm -f /tmp/mc litestream
```

All directory-replication databases replicated independently to MinIO, even with concurrent writes, and no path collisions were observed.

File: cmd/litestream/main.go [hunk 1]
```diff
@@ -440,9 +454,12 @@ type CompactionLevelConfig struct {
 	Interval time.Duration `yaml:"interval"`
 }
 
-// DBConfig represents the configuration for a single database.
+// DBConfig represents the configuration for a single database or directory of databases.
 type DBConfig struct {
 	Path               string         `yaml:"path"`
+	Directory          string         `yaml:"directory"` // Directory to scan for databases
```
Comment (benbjohnson) [inline cmd/litestream/main.go]: I'd shorten this down to `Dir` and then `dir` for the config. Not a big deal, just personal preference.

File: cmd/litestream/main.go [hunk 2]
```diff
@@ -512,6 +529,200 @@ func NewDBFromConfig(dbc *DBConfig) (*litestream.DB, error) {
...
+	dirPath, err := expand(dbc.Directory)
+	if err != nil {
+		return nil, err
+	}
+
+	// Default pattern if not specified
+	pattern := dbc.Pattern
+	if pattern == "" {
+		pattern = "*.db"
+	}
```
Comment (benbjohnson) [inline cmd/litestream/main.go]: There's not really a default extension for SQLite databases. Can you have it error if someone specifies a directory but not a pattern?

File: cmd/litestream/main.go [hunk 3]
```diff
@@ -512,6 +529,200 @@ func NewDBFromConfig(dbc *DBConfig) (*litestream.DB, error) {
...
+		// Calculate relative path from directory root
+		relPath, err := filepath.Rel(dirPath, dbPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to calculate relative path for %s: %w", dbPath, err)
+		}
+
+		// Create a copy of the config for each database
+		dbConfigCopy := *dbc
+		dbConfigCopy.Path = dbPath
+		dbConfigCopy.Directory = "" // Clear directory field for individual DB
```
Comment (benbjohnson) [inline cmd/litestream/main.go]: Maybe clear `Pattern` and `Recursive` too just for clarity.
